### PR TITLE
Add support for context menus (#139)

### LINKED ISSUE #139: Add context menu support
We've been working on a context menu support for GLSP and thought it might be useful to push this to Sprotty instead. What do you think?

The idea is that there is a generic `IContextMenuService` in Sprotty and depending on the context in which it is used (inside Theia, plain react app, or even Eclipse RCP) we bind a different implementation to this service that passes on the job of rendering the context menu to the "native" context menu implementation.

```typescript
export interface IContextMenuService {
    show(items: MenuItem[], anchor: Anchor, onHide?: () => void): void;
}
```

The context menu providers in Sprotty could look like below (so they specify all details about a menu item and define which Sprotty actions are going to be executed on click). To decide whether an item is enabled or not, it can use the state of the model (e.g. selection as below):

```typescript
@injectable()
export class DeleteContextMenuProvider implements IContextMenuProvider {
    getActions(root: Readonly<SModelRoot>, lastMousePosition?: Point): Promise<MenuItem[]> {
        const selectedElements = Array.from(root.index.all().filter(isSelected).filter(isDeletable));
        return Promise.resolve([
            {
                id: "delete",
                label: "Delete",
                sortString: "t",
                group: "a",
                actions: [new DeleteElementAction(selectedElements.map(e => e.id))],
                isEnabled: () => selectedElements.length > 0,
                isVisible: () => true,
                isToggled: () => false
            }
        ]);
    }
}
```

So far we've created an implementation for Theia based on Theia's context menu support (see [our work in progress implementation](https://github.com/eclipsesource/graphical-lsp/commit/fa14b76a8d6c14d7f73f08169297bb942bec0b0f#diff-0a363f67bcc2fb96de5ef8a2a5cf3fdb)). Essentially it registers the commands, menu actions, and submenus in Theia as well as a special command handler that passes on the Sprotty actions to the action dispatcher. It currently supports grouping, submenus, icons, sorting, enablement, toggle, as well as putting menu items into a parent that some other provider defines.

This is how it looks like in action:
Peek 2019-10-31 18-16

Before finalizing this change, I was wondering whether I should rather put it into Sprotty first and then do the GLSP integration (so that menu items can be provided also from a server). The work in progress implementation in GLSP is here: https://github.com/eclipsesource/graphical-lsp/commit/fa14b76a8d6c14d7f73f08169297bb942bec0b0f

Please let me know what you think! I'm happy to add this to Sprotty instead of GLSP only.

Happy Halloween! ;)

### PR DESCRIPTION
**Pull request by planger**
Fixes #139

### CODE REVIEW

**Comment by planger** [comment]
Please note that the implementation for Theia is provided in https://github.com/eclipse/sprotty-theia/pull/47

Unfortunately, these changes are a bit tedious to test as this change spans across sprotty and sprotty-theia, and then needs a Theia-based app that uses these changes. I've used https://github.com/eclipsesource/graphical-lsp/tree/context-menu-support for testing, but again it requires GLSP.

Is there a up-to-date Theia app into which I could integrate this so you can test it more easily (via yarn link)?

File: src/base/types.ts [hunk 1]
```diff
@@ -67,5 +67,9 @@ export const TYPES = {
     ICommandPaletteActionProviderRegistry: Symbol('ICommandPaletteActionProviderRegistry'),
     ICommandPaletteActionProvider: Symbol('ICommandPaletteActionProvider'),
     IEditLabelValidator: Symbol('IEditLabelValidator'),
-    IEditLabelValidationDecorator: Symbol('IEditLabelValidationDecorator')
+    IEditLabelValidationDecorator: Symbol('IEditLabelValidationDecorator'),
```
Comment (JanKoehnlein) [inline src/base/types.ts]: Minor comment here: Could we keep the alphabetic order (ignoring leading `I`)

Comment (JanKoehnlein) [inline src/base/types.ts]: Minor comment here: Could we keep the alphabetic order (ignoring leading I)

File: src/features/context-menu/menu-providers.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import { isSelected } from "../select/model";
+
+export interface IContextMenuItemProvider {
+    getItems(root: Readonly<SModelRoot>, lastMousePosition?: Point): Promise<LabeledAction[]>;
+}
+
+@injectable()
+export class ContextMenuProviderRegistry implements IContextMenuItemProvider {
+
+    constructor(@multiInject(TYPES.IContextMenuProvider) @optional() protected menuProviders: IContextMenuItemProvider[] = []) { }
```
Comment (JanKoehnlein) [inline src/features/context-menu/menu-providers.ts]: Shouldn't that symbol be called `IContextMenuItemProvider` then?

Comment (planger) [reply src/features/context-menu/menu-providers.ts]: Oh, sure... good catch!

File: src/base/types.ts (hunk 1 above)
Comment (planger) [reply src/base/types.ts]: Yes, I'll sort those in a sec.

**Comment by planger** [comment]
> As there are no tests, could you add a sample context menu to one of the examples?

Currently we only have an implementation of the `IContextMenuService` for Theia. So the example would have to be Theia-based at the moment. Is there a public example that I could use for demonstrating this change together with https://github.com/eclipse/sprotty-theia/pull/47? I'm happy to integrate it in such an example.

Alternatively, I could implement the context menu service for plain HTML pages, but that'd require more effort and I'm not sure how useful this would be for real applications, as you'd typically would want to use the same context menu implementation that is also used in the surrounding app (e.g. react-based, angular-based, or whatever).

Currently, this is already integrated in the GLSP workflow example, which is fairly simple. Ignoring server-provided menu items, this would look as follows:
1. The Theia context menu service must be handed to the sprotty container:
https://github.com/eclipsesource/graphical-lsp/commit/8aaf5ffc713f265c6898b7969bbdae22da0fe4ba#diff-6327e0a9b5c1f4e0e6fc0f16300391aa
2. Context menu item providers have to be created:
https://github.com/eclipsesource/graphical-lsp/commit/8aaf5ffc713f265c6898b7969bbdae22da0fe4ba#diff-6f3ee8e876e151ba5cee7ce458b47132R65
3. Context menu item providers have to be bound:
https://github.com/eclipsesource/graphical-lsp/commit/8aaf5ffc713f265c6898b7969bbdae22da0fe4ba#diff-c9f7fdba61179497d2459d5b7848d871R89

File: src/base/actions/action.ts [hunk 3]
```diff
@@ -62,3 +62,13 @@ export function isResponseAction(object?: any): object is ResponseAction {
             && typeof((object as any)['responseId']) === 'string'
             && (object as any)['responseId'] !== '';
 }
+
+export class LabeledAction {
```
[the remark this answers is not in the document]
Comment (planger) [reply src/base/actions/action.ts]: Right, done.

File: src/features/context-menu/di.config.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { ContainerModule } from "inversify";
+
+import { IContextMenuService } from "./context-menu-service";
+import { ContextMenuProviderRegistry } from "./menu-providers";
+import { ContextMenuMouseListener } from "./mouse-listener";
+import { TYPES } from "../../base/types";
+
+const contextMenuModule = new ContainerModule(bind => {
```
[the remark this answers is not in the document]
Comment (planger) [reply src/features/context-menu/di.config.ts]: Sure, I added it.

File: src/features/context-menu/context-menu-service.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    /** String indicating the grouping (separators). Items with equal group will be in the same group. */
+    readonly group?: string;
+    /**
+     * The optional parent id can be used to add this element as a child of another element provided by anohter menu provider.
+     * The `parentId` must be fully qualified in the form of `a.b.c`, whereas `a`, `b` and `c` are referring to the IDs of other elements.
+     * Note that this attribute will only be considered for root items of a provider and not for children of provided items.
+     */
+    readonly parentId?: string;
+    /** Function determining whether the element is enabled. */
+    readonly isEnabled?: () => boolean | boolean;
```
[the remark this answers is not in the document]
Comment (planger) [reply src/features/context-menu/context-menu-service.ts]: Oh, good catch, I changed it to `boolean | Promise<boolean>`.

**Comment by planger** [comment]
@spoenemann I've tried to integrate a context menu to npm-dependency-graph, but once I upgrade to the latest next versions of sprotty, sprotty-theia (and sprotty-elk), I get a couple of errors:
```
lerna ERR! execute callback with error
lerna ERR! Error: Command failed: yarn run prepare
lerna ERR! error Command failed with exit code 2.
lerna ERR! error Command failed with exit code 2.
lerna ERR! 
lerna ERR! yarn run v1.21.1
lerna ERR! $ yarn run clean && yarn run build
lerna ERR! $ rimraf lib
lerna ERR! $ tsc
lerna ERR! src/browser/graph/graph-layout.ts(18,15): error TS2416: Property 'graphOptions' in type 'DepGraphLayoutConfigurator' is not assignable to the same property in base type 'DefaultLayoutConfigurator'.
lerna ERR!   Type '(sgraph: SGraphSchema, index: SModelIndex<SModelElementSchema>) => LayoutOptions' is not assignable to type '(sgraph: SGraphSchema, index: SModelIndex<SModelElementSchema>) => LayoutOptions | undefined'.
lerna ERR!     Types of parameters 'index' and 'index' are incompatible.
lerna ERR!       Type 'import("/home/philip/Git/OpenSource/npm-dependency-graph/node_modules/sprotty/lib/base/model/smodel").SModelIndex<import("/home/philip/Git/OpenSource/npm-dependency-graph/node_modules/sprotty/lib/base/model/smodel").SModelElementSchema>' is not assignable to type 'import("/home/philip/Git/OpenSource/npm-dependency-graph/depgraph-navigator/node_modules/sprotty/lib/base/model/smodel").SModelIndex<import("/home/philip/Git/OpenSource/npm-dependency-graph/depgraph-navigator/node_modules/sprotty/lib/base/model/smodel").SModelElementSchema>'.
lerna ERR!         Types have separate declarations of a private property 'id2element'.
lerna ERR! src/browser/widget/diagram-commands.ts(49,37): error TS2352: Conversion of type 'ModelSource' to type 'DepGraphModelSource' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
lerna ERR!   Type 'ModelSource' is missing the following properties from type 'DepGraphModelSource': loadIndicator, graphGenerator, graphFilter, postConstruct, and 26 more.
lerna ERR! src/browser/widget/diagram-config.ts(35,38): error TS2345: Argument of type 'typeof TheiaKeyTool' is not assignable to parameter of type 'new (...args: any[]) => KeyTool'.
lerna ERR!   Type 'TheiaKeyTool' is not assignable to type 'KeyTool'.
lerna ERR!     Property 'keyListeners' is protected but type 'TheiaKeyTool' is not a class derived from 'KeyTool'.
lerna ERR! src/browser/widget/diagram-manager.ts(36,31): error TS2352: Conversion of type 'DiagramWidget' to type 'DepGraphWidget' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
lerna ERR!   Type 'DiagramWidget' is missing the following properties from type 'DepGraphWidget': searchBox, attached, diagramType
lerna ERR! src/browser/widget/diagram-manager.ts(38,9): error TS2322: Type 'DepGraphWidget' is not assignable to type 'DiagramWidget'.
lerna ERR!   Types of property 'modelSource' are incompatible.
lerna ERR!     Type 'DepGraphModelSource' is not assignable to type 'ModelSource'.
lerna ERR! src/browser/widget/diagram-widget.ts(25,9): error TS2416: Property 'modelSource' in type 'DepGraphWidget' is not assignable to the same property in base type 'DiagramWidget'.
lerna ERR!   Type 'DepGraphModelSource' is not assignable to type 'ModelSource'.
lerna ERR!     Property 'viewerOptions' is protected but type 'ModelSource' is not a class derived from 'ModelSource'.
lerna ERR! info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! 
lerna ERR!     at Promise.all.then.arr (/home/philip/Git/OpenSource/npm-dependency-graph/node_modules/execa/index.js:236:11)
lerna ERR!     at <anonymous>
{ Error: Command failed: yarn run prepare
error Command failed with exit code 2.
error Command failed with exit code 2.
```
I can look through those errors and try to fix them, but I probably won't be able to do that until the week of the 7th of Jan. Are you planning to upgrade npm-dependency-graph anyway? If yes, it'd be probably more efficient if I wait (since I don't really know the npm-dependency-graph).

Thanks!

**Comment by BjBe82** [comment]
I was interested in a context menu too so i spend some time to create a simple one using Sprotty UIExtension as well as the IContextMenuItemProvider and IContextMenuService api.

If you are interested i adapted the implementation for the mindmap example here: [https://github.com/BjBe82/sprotty/commit/c87eac0d9a75cc589b14067c981459bf54687423]

**Comment by planger** [comment]
> I was interested in a context menu too so i spend some time to create a simple one using Sprotty UIExtension as well as the IContextMenuItemProvider and IContextMenuService api.
> 
> If you are interested i adapted the implementation for the mindmap example here: [BjBe82@c87eac0](https://github.com/BjBe82/sprotty/commit/c87eac0d9a75cc589b14067c981459bf54687423)

Very interesting, thanks!

Imho it would be great to have a generic, Theia-independent implementation like yours in Sprotty for standalone deployments of Sprotty.

@spoenemann What do you think?

@BjBe82 Would you be interested in opening a PR where you put your context menu implementation (UI extension and ContextMenuService implementation) into an optional module in Sprotty and just use it in the mind-map example? The context menu provider should remain in the mindmap example.

Thanks for letting us know!

**Comment by spoenemann** [comment]
I'm not sure how useful it is in a stand-alone scenario. IMO a context menu is something for complex applications such as IDEs. "Normal" websites do not have custom context menus.

**Comment by JanKoehnlein** [comment]
Into the bargain, the surrounding framework often provides context menus and we should try to use that one to not look to alien, e.g. VS Code and Theia Desktop use the native context menu.

**Comment by BjBe82** [comment]
My intention was not to provide an extension of the sprotty api in that case because i agree with @spoenemann and @JanKoehnlein that advanced frameworks might provide a context menu implementation.

But for a small prototype i was interested in the sprotty context menu api so i created my own quite simple rendering. 

Here i just tried to provide a very lightweight example which could be added to the sprotty default examples if you are interested. Not saying that the rendering part is good or bad .... just to show that there is sprotty api around context menus and the fact that you discussed a “simple” example in this topic.

**Comment by spoenemann** [comment]
I'm fine with adding it to the mindmap example.

**Pull request merged**